service/storage_proxy.cc

```diff
@@ -2319,6 +2336,31 @@ future<> storage_proxy::send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s,
             allow_hints::no);
 }
 
+future<> storage_proxy::mutate_hint_from_scratch(frozen_mutation_and_schema fm_a_s) {
+    // FIXME: using 1h as infinite timeout. If a node is down, we should get an
+    // unavailable exception.
+    auto timeout = db::timeout_clock::now() + 1h;
+
+    if (!service::get_local_storage_service().cluster_supports_hinted_handoff_separate_connection()) {
+        return mutate({fm_a_s.fm.unfreeze(fm_a_s.s)}, db::consistency_level::ALL, timeout, nullptr, empty_service_permit());
+    }
+
+    std::array<mutation, 1> ms{std::move(fm_a_s.fm.unfreeze(fm_a_s.s))};
+
+    utils::latency_counter lc;
+    lc.start();
+    return mutate_prepare(std::move(ms), db::consistency_level::ALL, db::write_type::SIMPLE, empty_service_permit(),
+            [this] (const mutation& m,
+                db::consistency_level cl,
+                db::write_type type, service_permit permit) {
+        return create_hint_from_scratch_write_response_handler(m);
+    }).then([this, timeout] (std::vector<storage_proxy::unique_response_handler> ids) {
+        return mutate_begin(std::move(ids), db::consistency_level::ALL, timeout);
+    }).then_wrapped([this, p = shared_from_this(), lc] (future<> f) mutable {
+        return p->mutate_end(std::move(f), lc, _stats, nullptr);
+    });
```

Comment:
You basically copy paste `mutate_internal()` here. The way to extend `mutate_internal()` is to add an override for `storage_proxy::response_id_type storage_proxy::create_write_response_handler` that creates the kind of mutating_holder you want. Than you can use `mutate_internal()`  directly.

Reply:
Done - created an overload of `storage_proxy::create_write_response_handler` that takes `hint_wrapper` as the first parameter